### PR TITLE
Add binary support

### PR DESCRIPTION
**Pull request by arj03**
This adds support for transferring other data than js objects. A test is added that uses bipf, but this might as well be bamboo or gabby grove.

Furthermore I added a new section to the documentation that expands on how this works a bit better. This was after reading the original EBT paper.

The best way to review this is to disable whitespace changes as I reformatted some of the files.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -8,11 +8,11 @@
     "url": "git://github.com/dominictarr/epidemic-broadcast-trees.git"
   },
   "dependencies": {
-    "inherits": "^2.0.3",
```
Comment (arj03) [inline package.json:11]: this was never used

File: index.js [hunk 2]
```diff
@@ -49,8 +57,8 @@ module.exports = function (opts) {
     createStream: function (remote_id, version, client) {
       if(this.streams[remote_id])
         this.streams[remote_id].end(new Error('reconnected to peer'))
-      if(this.logging) console.error('EBT:conn', remote_id)
```
Comment (arj03) [inline index.js:52]: I changed these console.error to console.log as well.

File: stream.js [hunk 3]
```diff
@@ -1,124 +1,126 @@
...
-
-EBTStream.prototype.clock = function (clock) {
-  this.peer.state =
-    events.peerClock(this.peer.state, {id: this.remote, value: clock, ts: timestamp()})
-  this.paused = false
-  this.peer.update()
-  if(this.source) this.source.resume()
-}
-
-EBTStream.prototype._validate = function (clock) {
```
Comment (arj03) [inline stream.js:34]: noop

**Comment by arj03** [comment]
Ready for review now. I'll leave the other changes from #32 until this is reviewed.

**Comment by arj03** [comment]
Dum di dum di dum. Picking a random person to review this. @cryptix?

**Comment by cryptix** [comment]
I will give this a close read.

File: events.js [hunk 4]
```diff
@@ -1,472 +1,488 @@
...
-exports.note = opts.note
-exports.getReceive = opts.getReceive
-exports.getReplicate = opts.getReplicate
-exports.getSequence = opts.getSequence
-
-function isEmpty (o) {
-  for(var k in o) return false
-  return true
-}
+  function isEmpty (o) {
```
Comment (cryptix) [inline events.js:15]: while we are at cleaning up: do all these `function`s need to be scoped/closured  inside the `module.exports = (opts) => ...` thing?

File: events.js [hunk 5]
```diff
@@ -1,472 +1,488 @@
 'use strict'
 
 module.exports = function (opts) {
+  var exports = {}
```
Comment (cryptix) [inline events.js:4]: having the global `module.exports` and this variable is confusing. This var is the thing that get's returned by this func, right? Maybe `eventSomething`..? Processor or Actor.

side node; we have this un-named init thing for this sub-module and more exported funcs with names (clock, connect, etc.)?... :cold_sweat:

File: events.js [hunk 6]
```diff
@@ -1,472 +1,488 @@
...
+    return state
   }
 
-  return state
-}
+  exports.connect = function (state, ev) {
+    if(state.peers[ev.id]) throw new Error('already connected to peer:'+ev.id)
+    if(typeof ev.client != 'boolean') throw new Error('connect.client must be boolean')
+
+    //  if(isBlocked(state, state.id, ev.id)) return state
```
Comment (cryptix) [inline events.js:123]: commented out without a comment => remove?

File: events.js [hunk 7]
```diff
@@ -1,472 +1,488 @@
...
+  //and if so put this into lazy mode.
+  exports.notes = function (state, ev) {
+    //update replicating modes
+    var clock = ev.value
 
-  //support sending clocks inside a thing with additional properties.
-  //this is to allow room for backwards compatible upgrades.
-  if(isObject(ev.value.clock))
-    clock = ev.value.clock
+    //support sending clocks inside a thing with additional properties.
```
Comment (cryptix) [inline events.js:336]: this is about the protocol, right? IMHO deserves a place in the README

**Review by cryptix — APPROVED** [review]
Some style and clarity nit-picks on the Readme but mostly SGTM!

First time I read the event code, since the re-indent and moving around I was tempted to ask some opinionated questions but mostly style and naming things. Butt also SGTM.

File: README.md [hunk 8]
```diff
@@ -185,30 +206,73 @@ Read only access is okay, but updating should only be done via ebt methods.
...
+used for other things too)
+
+`<seq>` is an positive integer or zero. -1 is used to represent if the
+are explicitly not replicating that feed.
+
+`<msg>` is a message where `opts.isMsg(id) === true`.
+
+## Replication overview
+
+The state of other peers are stored outside this module in a lossy
```
[the remark this answers is not in the document]
Comment (arj03) [reply README.md:221]: I reworked that section a bit. Thanks

File: README.md [hunk 9]
```diff
@@ -185,30 +206,73 @@ Read only access is okay, but updating should only be done via ebt methods.
...
+
+Notes (aka the vector clock) are stored as { feed: (seq === -1 ? -1 :
+seq << 1 | !rx) } (= * 2 + 1?). The sequence can be extracted using
+`getSequence` and rx/tx using `getReceive` (is even). -1 means do not
+replicate.
+
+When peers connect, the server (that received the request) is expected
+to send their vector clock (notes) first. It should use a local cache
+as the last known status of the client. The notes will only contain
+feeds where the sequence is different from the other end. This ensures
```
[the remark this answers is not in the document]
Comment (arj03) [reply README.md:232]: :+1:

File: README.md [hunk 10]
```diff
@@ -185,30 +206,73 @@ Read only access is okay, but updating should only be done via ebt methods.
...
+that the vectors clocks sent are as small as possible.
+
+When connecting to multiple nodes, only request new messages using rx
+for a feed from one of the nodes. See `test/multiple.js`.
```
[the remark this answers is not in the document]
Comment (arj03) [reply README.md:236]: Good idea

File: README.md [hunk 11]
```diff
@@ -185,30 +206,73 @@ Read only access is okay, but updating should only be done via ebt methods.
...
+signal of what feeds to replicate. EBT won't connect to someone that
+has been blocked. It will not send messages of a peer (including self)
+to another peer if the first peer blocks the second.
+
+The tests are very readable because they use a simulator where a trace
+of the run is saved and pretty printed. See `test/two.js` for a good
+example.
+
+TODO:
+ - is there a way to reset a remote clock? In the case where one nukes
```
[the remark this answers is not in the document]
Comment (arj03) [reply README.md:248]: Great idea. This was the last items on my TODO list :D

File: events.js (hunk 4 above)
Comment (arj03) [reply events.js:15]: If they are only used in the file I guess that is fine?

Comment (arj03) [reply events.js:15]: Hmm yeah, it is a bit confusing in the other files. Will make a note of that. Thanks

File: events.js (hunk 5 above)
Comment (arj03) [reply events.js:4]: Agree the opts is quite confusing. One of the other changes I have queued up in removing v2. And when we do this we don't really need this opts.

File: events.js (hunk 6 above)
Comment (arj03) [reply events.js:123]: Agree, that is a bit strange. It was already there, I didn't actually see it before you pointed it out.

File: events.js (hunk 7 above)
Comment (arj03) [reply events.js:336]: Agree. I've created an issue for that. This was also already there.

**Comment by arj03** [comment]
Thanks a lot for taking the time to review this @cryptix. It is not the easiest module to get started on, but what I wanted to do here is to lay the foundation for better documenting for how the system works, as to make it easier for other languages to implement the protocol. I realize we probably need a lot more, but hey, its a start :-)

Just today I read another article on Brisa which is another EBT system that references the original plumtree article. So I'll try and add some more context to this module because the protocol is actually REALLY interesting in that I havn't been able to find a scientific article that describes a system where you are basically creating a tree per feed with a large degree of overlap between what peers are storing the feeds.

**Comment by christianbundy** [comment]
This is awesome!

**Comment by arj03** [comment]
Thanks @christianbundy. After the all the changes from #32 have been merged, I think it would be nice with a ESlint on this to see what it can dig up, if you are up for that :-)

**Pull request merged**